src/pages/library/datasetBuilder/CohortEditor.ts

```diff
@@ -0,0 +1,442 @@
+                  div({
+                    style: { marginLeft: narrowMargin, flexGrow: 1, borderTop: `1px solid ${colors.dark(0.35)}` },
+                  }),
+                ]),
+              ]),
+            Utils.toIndexPairs(cohort.criteriaGroups)
+          ),
+        ]),
+  ]);
+};
+
+const editorBackgroundColor = colors.light(0.7);
+
+type CohortEditorContentsProps = {
+  updateCohort: CohortUpdater;
+  cohort: Cohort;
+  datasetDetails: DatasetResponse;
+  onStateChange: OnStateChangeHandler;
+};
+const CohortEditorContents: React.FC<CohortEditorContentsProps> = (props) => {
+  const { updateCohort, cohort, datasetDetails, onStateChange } = props;
+  return div(
+    {
+      style: { padding: `${PAGE_PADDING_HEIGHT}rem ${PAGE_PADDING_WIDTH}rem` },
+    },
+    [
+      h2({ style: { display: 'flex', alignItems: 'center' } }, [
+        h(
+          Link,
+          {
+            onClick: () => {
+              onStateChange(new HomepageState());
+            },
+            'aria-label': 'cancel',
+          },
+          [icon('left-circle-filled', { size: 32 })]
+        ),
+        div({ style: { marginLeft: 15 } }, [cohort.name]),
+      ]),
+      h3(['To be included in the cohort, participants...']),
+      div({ style: { display: 'flow' } }, [
+        h(CohortGroups, {
+          key: cohort.name,
+          datasetDetails,
+          cohort,
+          updateCohort,
+        }),
+        h(
+          ButtonOutline,
+          {
+            style: { marginTop: wideMargin },
+            onClick: (e) => {
+              updateCohort(_.set(`criteriaGroups.${cohort.criteriaGroups.length}`, newCriteriaGroup()));
+              // Lose button focus, since button moves out from under the user's cursor.
+              e.currentTarget.blur();
+            },
+          },
+          ['Add group']
+        ),
+      ]),
+    ]
+  );
+};
+
+interface CohortEditorProps {
+  onStateChange: OnStateChangeHandler;
+  datasetDetails: DatasetResponse;
+  originalCohort: Cohort;
+}
+
+type CohortUpdater = (updater: (cohort: Cohort) => Cohort) => void;
+
+export const CohortEditor: React.FC<CohortEditorProps> = (props) => {
+  const { onStateChange, datasetDetails, originalCohort } = props;
+  const [cohort, setCohort] = useState(originalCohort);
+  const updateCohort: CohortUpdater = (updateCohort: (Cohort) => Cohort) => {
+    _.flow(updateCohort, setCohort)(cohort);
+  };
+
+  return h(Fragment, [
+    h(CohortEditorContents, { updateCohort, cohort, datasetDetails, onStateChange }),
+    // add div to cover page to footer
+    div(
+      {
+        style: {
+          display: 'flex',
+          backgroundColor: editorBackgroundColor,
+          alignItems: 'end',
+          flexDirection: 'row-reverse',
+          padding: wideMargin,
+        },
+      },
+      [
+        h(
+          ButtonPrimary,
+          {
+            onClick: () => {
+              const cohorts: Cohort[] = datasetBuilderCohorts.get();
```

Comment:
I'm curious why an Atom-backed state mechanic is required here... wouldn't having the consumer/composer of CohortEditor responding to onStateChanged event be sufficient?

Reply:
It's written this way because the original code used different pages for this UI. Now that it's on a single page, we can move this into component state instead. I'll create a JIRA ticket for that work.